### PR TITLE
AP-4054: generate and attach results CSV file

### PR DESCRIPTION
**Pull request by jsugarman**
## What
Generate and attach CSV file for each bulk_submission

[Link to story](https://dsdmoj.atlassian.net/browse/AP-4054)

Generate a downloadable file for case workers to download as part of the nightly processing
of any bulk submission file uploads.

Screenshot 2023-05-26 at 07 40 59

Image of a downloaded file of sandbox users in Excel (can also be opened in numbers, Mac OS)

Screenshot 2023-05-25 at 17 27 30

## Checklist

Before you ask people to review this PR:

- Tests and linters should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

### CODE REVIEW

**Comment by jsugarman** [comment]
> Looks good as a first pass 👍 Was initially unsure whether we needed so many statuses for bulk submissions but have had a look and can see the logic for each. Would be good to get code coverage back up.

It's so we can track where problems have occurred in the flow more easily. Errors could occur during any job for any submission potentially and at any phase. The status attribute plus the isolation of the jobs, a practice which helps with error handling and separation of queueing and retry logic, means we can customise specific jobs/workers as and when we find problems. Plus no single failure will cause the overall bulk submission to fail (or that is the idea 😬)

**Review by kmahern — APPROVED** [review]
LGTM 👍

File: spec/lib/govuk_component/tag_helpers_spec.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      subject(:markup) do
+        helper.govuk_status_tag("ready")
+      end
+
+      it 'adds tag with color class, prepended by govuk class' do
+        expect(markup).to have_tag(:strong, with: { class: 'govuk-tag govuk-tag--green' })
+      end
+    end
+
+    context 'without unregistered status' do
```
Comment (jsugarman) [inline spec/lib/govuk_component/tag_helpers_spec.rb:27]:
```suggestion
    context 'with status without colour mapping' do
```

File: spec/services/bulk_submission_service_spec.rb [hunk 2]
```diff
@@ -78,9 +85,56 @@
...
+      end
+
+      it "enqueues HmrcInterfaceBulkSubmissionWorker job for bulk_submission", type: :worker do
+        expect { call }
+          .to change(HmrcInterfaceBulkSubmissionWorker, :jobs)
+          .from([])
+          .to(
+            [
+              hash_including(
+                "retry" => true,
```
Comment (jsugarman) [inline spec/services/bulk_submission_service_spec.rb:103]: 🤔 perhaps we should be limiting this to 5 or some number lower than that for dependant workers

File: spec/system/bulk_submission_index_spec.rb [hunk 3]
```diff
@@ -29,17 +29,104 @@
...
+        end
+      end
+    end
+
+    context "with an existing ready bulk_submission" do
+      before do
+        create(:bulk_submission, :with_original_file, :with_result_file, :ready)
+      end
+
+      it "user can remove them" do
```
Comment (jsugarman) [inline spec/system/bulk_submission_index_spec.rb:78]: Add "user can download them" test

File: app/views/bulk_submissions/index.html.erb [hunk 4]
```diff
@@ -7,23 +7,33 @@
     <%= govuk_table do |table|
       table.with_head do |head|
         head.with_row do |row|
-          row.with_cell(text: "Filename", header: true) if Rails.env.development?
           row.with_cell(text: t(".created_at"), header: true)
           row.with_cell(text: t(".expires_at"), header: true)
+          row.with_cell(text: t(".filename"), header: true)
           row.with_cell(text: t(".status"), header: true)
           row.with_cell(text: t(".action"), header: true)
+          row.with_cell(text: "")
```
Comment (naseberry) [inline app/views/bulk_submissions/index.html.erb:15]: Accessibility/Improvement: 

Columns can be made to stretch across multiple rows with `colspan` parameter, therefore, you could remove the empty head cell and apply the `colspan` parameter to the previous sibling.

for example: `row.with_cell(text: t(".action"), header: true, colspan: 2, scope: 'colgroup')`

File: app/models/bulk_submission.rb [hunk 5]
```diff
@@ -3,6 +3,25 @@ class BulkSubmission < ApplicationRecord
...
+  has_one_attached :result_file
+
+  has_status :pending,
+             :preparing,
+             :prepared,
+             :processing,
+             :completed,
+             :writing,
+             :exhausted,
+             :ready
```
Comment (naseberry) [inline app/models/bulk_submission.rb:21]: status seems to be missing `failed`

File: app/services/bulk_submission_service.rb [hunk 6]
```diff
@@ -48,4 +49,9 @@ def file_records
   def file_contents
     @file_contents ||= bulk_submission.original_file.download
   end
+
+  # approx 7 seconds for HMRC interface to return a result for an existing person
+  def delay
+    bulk_submission.submissions.count * 7
```
Comment (naseberry) [inline app/services/bulk_submission_service.rb:55]: discussion: have we considered network delays, do we have a retry method?

File: app/views/bulk_submissions/index.html.erb [hunk 7]
```diff
@@ -1,29 +1,39 @@
 <div class="govuk-grid-row">
-  <div class="govuk-grid-column-two-thirds">
+  <div class="govuk-grid-column-three-thirds">
```
Comment (colinbruce) [inline app/views/bulk_submissions/index.html.erb:2]: Isn't three thirds the whole width?

File: app/views/bulk_submissions/index.html.erb [hunk 8]
```diff
@@ -1,29 +1,39 @@
...
-              button_to('Remove', bulk_submission_path(bulk_submission.id), method: :delete)
+              if bulk_submission.pending?
+                button_to('Cancel', bulk_submission_path(bulk_submission.id), method: :delete)
+              elsif bulk_submission.ready?
+                link_to "Download", rails_blob_path(bulk_submission.result_file.blob, disposition: 'attachment')
+              end
+            end
+            row.with_cell do
+              if bulk_submission.ready?
+                button_to('Remove', bulk_submission_path(bulk_submission.id), method: :delete)
```
Comment (naseberry) [inline app/views/bulk_submissions/index.html.erb:35]: Accessibility:

provide context for the action `Remove <span class="govuk-visually-hidden"> filename</span>`

File: config/application.rb [hunk 9]
```diff
@@ -18,6 +17,9 @@
 # you've limited to :test, :development, or :production.
 Bundler.require(*Rails.groups)
 
+# Custom railties that are not gems can be required here
+require_relative '../lib/govuk_component'
```
Comment (naseberry) [inline config/application.rb:21]: 👍

**Comment by naseberry** [comment]
Great first pass, a few bits to tweak 👍

File: app/models/bulk_submission.rb (hunk 5 above)
Comment (jsugarman) [reply app/models/bulk_submission.rb:21]: that status is only for submissions, not bulk_submissions, and is one returned by hmrcinterface

File: app/services/bulk_submission_service.rb (hunk 6 above)
Comment (jsugarman) [reply app/services/bulk_submission_service.rb:55]: lets see how it plays out and we can adjust if we are finding problems

File: app/views/bulk_submissions/index.html.erb (hunk 7 above)
Comment (jsugarman) [reply app/views/bulk_submissions/index.html.erb:2]: don't know my govuk css classes, is that `govuk-grid-column-full` or `govuk-grid-column`?

**Review by naseberry — APPROVED** [review]

**Pull request merged**